routing/router.go

```diff
@@ -147,6 +147,29 @@ type PaymentAttemptDispatcher interface {
 		<-chan *htlcswitch.PaymentResult, error)
 }
 
+// PaymentSessionSource is an interface that defines a source for the router to
+// retrive new payment sessions.
+type PaymentSessionSource interface {
+
+	// NewPaymentSession creates a new payment session that will produce
+	// routes to the given target. An optional set of routing hints can be
+	// provided in order to populate additional edges to explore when
+	// finding a path to the payment's destination.
+	NewPaymentSession(routeHints [][]zpay32.HopHint,
+		target route.Vertex) (PaymentSession, error)
+
+	// NewPaymentSessionForRoute creates a new paymentSession instance that
+	// is just used for failure reporting to missioncontrol, and will only
+	// attempt the given route.
+	NewPaymentSessionForRoute(preBuiltRoute *route.Route) PaymentSession
+
+	// NewPaymentSessionEmpty creates a new paymentSession instance that is
+	// empty, and will be exhausted immediately. Used for failure reporting
+	// to missioncontrol for resumed payment we don't want to make more
+	// attempts for.
+	NewPaymentSessionEmpty() PaymentSession
+}
```

Comment:
I am not sure if this is the right cut to make for testing. The way I see it, `router` creates payment sessions, either on demand or after restart. Then `sendPayment` is operating on a payment session to execute the payment. Ideally (imo) sendPayment is a method of payment session even.

For unit tests, I would expect some tests that exercise the behaviour of `sendPayment` by passing in a mock session and then there are tests that assert that `router` behaves correctly and instantiates the right payment sessions.

Reply:
AFAICT, the current split is that the payment session represents our internal routing state, therefore under this mode, I think actually sending payments would be counted as an "external" factor.

Reply:
This is designed around letting payment session being our source for getting new routes. With the new ´paymenLifecycle` struct, we move closer to @joostjager's suggestion, where we can mock out the router eventually.